### PR TITLE
Np 3496 configurable cristinclient

### PR DESCRIPTION
**Pull request by svenngjeto**

### CODE REVIEW

File: cristin-commons/src/main/java/no/unit/nva/cristin/common/client/ApiClient.java [hunk 1]
```diff
@@ -149,4 +161,61 @@ private boolean responseIsFailure(int statusCode) {
...
+        return effortCount;
+    }
+
+    public Integer calculateFirstRecord(Map<String, String> requestQueryParams) {
+        final int page = Integer.parseInt(requestQueryParams.get(PAGE));
+        final int pageSize = Integer.parseInt(requestQueryParams.get(NUMBER_OF_RESULTS));
+        return (page - 1) * pageSize + 1;
+    }
+
+    public int getCount(HttpResponse<String> response, List<Organization> organizations) {
```
Comment (torarnet) [inline cristin-commons/src/main/java/no/unit/nva/cristin/common/client/ApiClient.java:211]: Can this be more generic so we can use it for more than organization?

File: cristin-commons/src/main/java/no/unit/nva/cristin/model/Constants.java [hunk 2]
```diff
@@ -22,14 +22,16 @@
...
     public static final String PROJECT_LOOKUP_CONTEXT_URL =
         "https://bibsysdev.github.io/src/project-context.json";
-    public static final String CRISTIN_API_HOST = ENVIRONMENT.readEnvOpt("CRISTIN_API_HOST")
-        .orElse("api.cristin.no");
-    public static final String CRISTIN_API_BASE = CRISTIN_API_HOST + "/v2";
-    // TODO: Find better way to use Cristin test instance
-    public static final String CRISTIN_TEST_API_HOST = "api.cristin-test.uio.no";
-    public static final String CRISTIN_TEST_API_BASE = CRISTIN_TEST_API_HOST + "/v2";
-
+//    public static final String CRISTIN_API_HOST = ENVIRONMENT.readEnvOpt("CRISTIN_API_HOST")
```
Comment (torarnet) [inline cristin-commons/src/main/java/no/unit/nva/cristin/model/Constants.java:25]: Comments?

File: cristin-commons/src/main/java/no/unit/nva/utils/UriUtils.java [hunk 3]
```diff
@@ -77,10 +78,15 @@ public static URI createIdUriFromParams(Map<String, String> requestQueryParams,
...
-        URI queryUri = new UriWrapper(HTTPS, CRISTIN_API_BASE)
+        URI queryUri = new UriWrapper(CRISTIN_API_URL)
                 .addChild(path)
                 .addQueryParameters(cristinRequestQueryParams)
                 .getUri();
         return queryUri;
     }
+
+    public static  URI getNvaApiBaseUri() {
+        return new UriWrapper(HTTPS, DOMAIN_NAME).addChild(BASE_PATH).addChild(ORGANIZATION_PATH).getUri();
```
Comment (torarnet) [inline cristin-commons/src/main/java/no/unit/nva/utils/UriUtils.java:89]: Is `getNvaApiBaseUri` correct name when adding child `ORGANIZATION_PATH`

File: cristin-project/src/test/java/no/unit/nva/cristin/projects/CristinApiClientTest.java [hunk 4]
```diff
@@ -41,9 +41,9 @@
 public class CristinApiClientTest {
 
     private static final String QUERY_CRISTIN_PROJECTS_EXAMPLE_URI =
-        "https://api.cristin.no/v2/projects/?lang=nb&page=1&per_page=5&title=reindeer";
+        "https://api.cristin-test.uio.no/v2/projects?per_page=5&page=1&lang=nb&title=reindeer";
     private static final String CRISTIN_API_GRANT_ID_SEARCH_EXAMPLE_URI =
-        "https://api.cristin.no/v2/projects/?lang=nb&page=1&per_page=5&project_code=1234567";
+        "https://api.cristin-test.uio.no/v2/projects?per_page=5&project_code=1234567&page=1&lang=nb";
```
Comment (torarnet) [inline cristin-project/src/test/java/no/unit/nva/cristin/projects/CristinApiClientTest.java:46]: Will this param ordering always be correct?

**Review by torarnet — COMMENTED** [review]
See comments

File: cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java [hunk 5]
```diff
@@ -2,77 +2,141 @@
...
+                CRISTIN_PER_PAGE_PARAM, requestQueryParams.get(NUMBER_OF_RESULTS));
+    }
+
+    private SearchResponse<Organization> updateSearchResponseMetadata(
+            SearchResponse<Organization> searchResponse,
+            Map<String, String> requestQueryParams,
+            long timeUsed) {
+        searchResponse.setContext(ORGANIZATION_CONTEXT);
+        searchResponse.setId(createIdUriFromParams(requestQueryParams, ORGANIZATION_PATH));
+        if (searchResponse.getSize() > 0) {
```
Comment (brinxmat) [inline cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java:108]: isNotEmpty

File: cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java [hunk 6]
```diff
@@ -2,77 +2,141 @@
...
+        searchResponse.setProcessingTime(timeUsed);
+        return searchResponse;
+    }
+
+    private URI previousResult(URI baseUri, Map<String, String> requestQueryParams, int totalSize) {
+        int firstPage = Integer.parseInt(requestQueryParams.get(PAGE)) - 1;
+        if (firstPage > 0 && totalSize > 0) {
+            Map<String, String> nextMap = new ConcurrentHashMap<>(requestQueryParams);
+            nextMap.put(PAGE, Integer.toString(firstPage));
+            return new UriWrapper(baseUri).addQueryParameters(nextMap).getUri();
```
Comment (brinxmat) [inline cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java:124]: Extract method, which covers both previous and next result URI.

Comment (brinxmat) [reply cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java:124]: Use ? : syntax to provide single return

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: cristin-project/src/test/java/no/unit/nva/cristin/projects/CristinApiClientTest.java (hunk 4 above)
Comment (svenngjeto) [reply cristin-project/src/test/java/no/unit/nva/cristin/projects/CristinApiClientTest.java:46]: It stays the same in a deterministic way - if it is in correct logical order is another question...  Maybe we should have 
 something in nva.testutils to decompose an compare URIs

**Review by torarnet — APPROVED** [review]

**Review by brinxmat — APPROVED** [review]

**Pull request merged**